Changelog 128 (#90)
intellij-rust/intellij-rust.github.io: _posts/2020-08-06-changelog-128.markdown

---
layout: post
title: "IntelliJ Rust Changelog #128"
date: 2020-08-06 12:00:00 +0300
---

## New Features

<!-- https://github.com/intellij-rust/intellij-rust/pull/5715 -->
* Improve `New project` dialog to support `cargo-generate` templates.
Now a WebAssembly project can be created using the corresponding template (by [@avrong])

    {% include gif-img.html path="/assets/posts/changelog-128/cargo-generate" w="700px" %}

<!-- https://github.com/intellij-rust/intellij-rust/pull/5451 -->

> * Add intention to import a complex path.

Reviewer: Maybe `qualified` is better than `complex`?

> Now you can add `use` item for any path under the caret (by [@Kobzol])

Reviewer: nit: `any` isn't correct. Only qualified paths have this intention

    {% include gif-img.html path="/assets/posts/changelog-128/import-path" w="500px" %}

* `Move` refactoring (<kbd>F6</kbd>) improvements:
    <!-- https://github.com/intellij-rust/intellij-rust/pull/5736 -->
    * Search and update changed references when moving a top-level item
        

>         {% include gif-img.html path="/assets/posts/changelog-128/move-update-references" w="600px" %}

Reviewer: Personally, I don't understand what the feature is

    <!-- https://github.com/intellij-rust/intellij-rust/pull/5484 -->
    * Take reexports into account when moving a single file

        {% include gif-img.html path="/assets/posts/changelog-128/move-reexports" w="600px" %}

<!-- https://github.com/intellij-rust/intellij-rust/pull/5758 -->
* Provide `true`/`false` completion for boolean values (by [@ozkriff]). N

    {% include gif-img.html path="/assets/posts/changelog-128/bool-completion" w="700px" %}

* Improve `Wrong number of type arguments` inspection
    <!-- https://github.com/intellij-rust/intellij-rust/pull/5811 -->
    * Support trait objects (by [@Kobzol])

    <!-- https://github.com/intellij-rust/intellij-rust/pull/5306 -->

>     * Provide `Add missing type arguments` fix (by [@Kobzol])

Reviewer: Let's add a note that the fix adds only type parameters and ignores lifetimes and associated types

        {% include gif-img.html path="/assets/posts/changelog-128/missing-type-arg" w="700px" %}
            
* Improve struct patterns completion:
    <!-- https://github.com/intellij-rust/intellij-rust/pull/5660 -->

>     * Suggest only struct fields in struct pattern completion (by [@Kobzol])

Reviewer: ```suggestion
    * Suggest only struct fields while completion in struct patterns (by [@Kobzol])
```

    
        {% include gif-img.html path="/assets/posts/changelog-128/struct-fields-comp" w="700px" %}

    <!-- https://github.com/intellij-rust/intellij-rust/pull/5825 -->
    * Completion for `..` syntax in struct patterns (by [@Kobzol])

<!-- https://github.com/intellij-rust/intellij-rust/pull/5808 -->

> * Better type inference for `associated_type_bounds` rustc nightly feature

Reviewer: nit: Oh, it's not "Better" because we didn't suppport this feature before this change

<!-- https://github.com/intellij-rust/intellij-rust/pull/5643 -->
* Add `Substitute associated type` intention (by [@Kobzol])

<!-- https://github.com/intellij-rust/intellij-rust/pull/5798 -->
* Add `Main function not found in crate` ([E0601](https://doc.rust-lang.org/error-index.html#E0601)) inspection for build scripts (by [@Stzx])

<!-- https://github.com/intellij-rust/intellij-rust/pull/5747 -->
* Fold external (non-project related) parts of stack traces (by [@Kobzol])

## Fixes

* Type inference fixes:
    <!-- https://github.com/intellij-rust/intellij-rust/pull/5841 -->
    * Infer type parameters from associated type bindings
    
    <!-- https://github.com/intellij-rust/intellij-rust/pull/5837 -->
    * Infer types of complex `?` expressions
    
    <!-- https://github.com/intellij-rust/intellij-rust/pull/5824 -->
    * Ignore currently unsupported negative `impl` blocks during type inference
    
    <!-- https://github.com/intellij-rust/intellij-rust/pull/5812 -->
    * Fix `self` substitution in bounds during type inference
    
    <!-- https://github.com/intellij-rust/intellij-rust/pull/5809 -->
    * Fix nested projections normalization
    
    

> * Macro expansion fixes:    

Reviewer: ```suggestion
* Macro expansion improvements:  
```

    <!-- https://github.com/intellij-rust/intellij-rust/pull/5862 -->
    * Fix `ArrayIndexOutOfBoundsException` during macro call arguments completion
    
    <!-- https://github.com/intellij-rust/intellij-rust/pull/5810 -->
    * Replace `IntellijRustDollarCrate` with a crate name in `Show macro expansion` action
    
    <!-- https://github.com/intellij-rust/intellij-rust/pull/5796 -->

>     * Resolve `$crate` via hygiene

Reviewer: What does it mean from user side?

    
    <!-- https://github.com/intellij-rust/intellij-rust/pull/5855 -->
    * Don't skip doc comments provided via macro definition
    
    <!-- https://github.com/intellij-rust/intellij-rust/pull/5777 -->
    * Fix groups substitution in macro expansion

<!-- https://github.com/intellij-rust/intellij-rust/pull/5857 -->
* Find and replace occurrences located outside a function in `Introduce constant/variable` refactoring (by [@Kobzol])

<!-- https://github.com/intellij-rust/intellij-rust/pull/5856 -->
* Properly check spelling in lifetimes. Now `'static` is not reported as a typo

<!-- https://github.com/intellij-rust/intellij-rust/pull/5853 -->
* Fix name resolution of private reexport with `crate::` path

<!-- https://github.com/intellij-rust/intellij-rust/pull/5849 -->
* Include `RUSTC` in environment when executing Cargo (by [@tamird])

<!-- https://github.com/intellij-rust/intellij-rust/pull/5842 -->

> * Don't show parameter hints for single-parameter closures

Reviewer: Let's add note about "smart" hint option to return the corresponding hints back if needed

<!-- https://github.com/intellij-rust/intellij-rust/pull/5838 -->
* Fix stdlib source paths on nightly. This fixes `Cargo project update failed` error.

<!-- https://github.com/intellij-rust/intellij-rust/pull/5822 -->
* Don't provide color highlighting in batch mode. It prevents empty items in `Inspection Result` tool window

<!-- https://github.com/intellij-rust/intellij-rust/pull/5820 -->
* Suppress `snake_case` naming lint for `#[no_mangle]` functions (by [@Kobzol])

<!-- https://github.com/intellij-rust/intellij-rust/pull/5819 -->
* Add `unsafe` modifier to functions [extracted](https://www.jetbrains.com/help/idea/extract-method.html) from unsafe functions (by [@Kobzol])

<!-- https://github.com/intellij-rust/intellij-rust/pull/5801 -->
* Fix navigation to derive items in 2020.2

<!-- https://github.com/intellij-rust/intellij-rust/pull/5799 -->
* Use correct language name for Toml color settings configurable

<!-- https://github.com/intellij-rust/intellij-rust/pull/5789 -->

> * Don't create checkboxes in inlay hint settings manually

Reviewer: The fix (from user side) is about proper redndering

<!-- https://github.com/intellij-rust/intellij-rust/pull/5779 -->
* Fix mod declarations with `#[path = "mod.rs"]` attribute

<!-- https://github.com/intellij-rust/intellij-rust/pull/5766 -->
* Attach module without a parent to a parent module in `Attach file to module` fix (by [@Kobzol])

<!-- https://github.com/intellij-rust/intellij-rust/pull/5661 -->

> * Prefer trait methods before private inherent methods in completion (by [@Kobzol])

Reviewer: Let's move this item higher since it's more important than, for example, https://github.com/intellij-rust/intellij-rust/pull/5799 or https://github.com/intellij-rust/intellij-rust/pull/5789

## Internal Improvements

<!-- https://github.com/intellij-rust/intellij-rust/pull/5840 -->

> * Allow triggering github workflow from UI

Reviewer: ```suggestion
* Allow triggering project GitHub workflows from UI
```
